internal/config/config.go

```diff
@@ -36,17 +38,23 @@ type Config struct {
 	Branch string
 	// JobRetryCount is the count of the number of times the job has been retried.
 	JobRetryCount int
+	// Env provides access to environment variables.
+	// It's public because many tests in other packages reference it (perhaps they should not).
+	Env env.Env
 	// errs is a map of environment variables name and the validation errors associated with them.
 	errs InvalidConfigError
 }
 
 // New wraps the readFromEnv and validate functions to create a new Config struct.
 // It returns Config struct and an InvalidConfigError if there is an invalid configuration.
-func New() (Config, error) {
-	c := Config{errs: InvalidConfigError{}}
+func New(env env.Env) (Config, error) {
+	c := Config{
+		Env:  env,
+		errs: InvalidConfigError{},
+	}
```

Comment:
It isn't quite clear to me whether this level of abstraction is needed for this project. 

This project currently use `Config` as the main injection point to set env during test, many test cases rely on explicitly setting `Config`, which doesn't seem to be troublesome. 

Using the new `env` parameter as the new injection point, we will probably have to refactor many test cases to make it obvious, otherwise I think most devs will continue to assume `Config` being the boundary. 

I can see benefits such as allowing parallel execution but I doubt if this is the goal? I probably missed some more contexts?

Reply:
I think it should _mostly_ be useful when testing the existing `main` implementation, and when testing the upcoming subcommands (`upload`, `bazel listen`, …) which also turn `(argv, envp)` into configuration.

Aside from that, a key thing with `bktec upload` is that a whole lot of common CI environment variables are checked for and, if present, mapped into `run_env`. For example `BUILDKITE_BUILD_ID` etc. These are (arguably) not really `bktec` _configuration_… `bktec` is probing the environment to find out the context that it's being executed within.

So for that reason, I think it's worthwhile to be able to swap in a fake env for testing.